ssl/ech/ech_internal.c

```diff
@@ -160,4 +227,1223 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+ */
+int ech_send_grease(SSL_CONNECTION *s, WPACKET *pkt)
+{
+    OSSL_HPKE_SUITE hpke_suite_in = OSSL_HPKE_SUITE_DEFAULT;
+    OSSL_HPKE_SUITE *hpke_suite_in_p = NULL;
+    OSSL_HPKE_SUITE hpke_suite = OSSL_HPKE_SUITE_DEFAULT;
+    size_t pp_at_start = 0, pp_at_end = 0;
+    size_t senderpub_len = OSSL_ECH_MAX_GREASE_PUB;
+    size_t cipher_len = 0, cipher_len_jitter = 0;
+    unsigned char cid, senderpub[OSSL_ECH_MAX_GREASE_PUB];
+    unsigned char cipher[OSSL_ECH_MAX_GREASE_CT];
+    unsigned char *pp = WPACKET_get_curr(pkt);
+
+    if (s == NULL)
+        return 0;
+    if (s->ssl.ctx == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    WPACKET_get_total_written(pkt, &pp_at_start);
+    /* randomly select cipher_len to be one of 144, 176, 208, 244 */
+    if (RAND_bytes_ex(s->ssl.ctx->libctx, &cid, 1,
+                      RAND_DRBG_STRENGTH) <= 0) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    cipher_len_jitter = cid % 4;
+    cipher_len = 144;
+    cipher_len += 32 * cipher_len_jitter;
+    /* generate a random (1 octet) client id */
+    if (RAND_bytes_ex(s->ssl.ctx->libctx, &cid, 1,
+                      RAND_DRBG_STRENGTH) <= 0) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    s->ext.ech.attempted_cid = cid;
+    hpke_suite_in_p = &hpke_suite;
+    if (s->ext.ech.grease_suite != NULL) {
+        if (OSSL_HPKE_str2suite(s->ext.ech.grease_suite, &hpke_suite_in) != 1) {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            return 0;
+        }
+        hpke_suite_in_p = &hpke_suite_in;
+    }
+    if (OSSL_HPKE_get_grease_value(hpke_suite_in_p, &hpke_suite,
+                                   senderpub, &senderpub_len,
+                                   cipher, cipher_len,
+                                   s->ssl.ctx->libctx, NULL) != 1) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (!WPACKET_put_bytes_u16(pkt, s->ext.ech.attempted_type)
+        || !WPACKET_start_sub_packet_u16(pkt)
+        || !WPACKET_put_bytes_u8(pkt, OSSL_ECH_OUTER_CH_TYPE)
+        || !WPACKET_put_bytes_u16(pkt, hpke_suite.kdf_id)
+        || !WPACKET_put_bytes_u16(pkt, hpke_suite.aead_id)
+        || !WPACKET_memcpy(pkt, &cid, 1)
+        || !WPACKET_sub_memcpy_u16(pkt, senderpub, senderpub_len)
+        || !WPACKET_sub_memcpy_u16(pkt, cipher, cipher_len)
+        || !WPACKET_close(pkt)
+        ) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /* record the ECH sent so we can re-tx same if we hit an HRR */
+    OPENSSL_free(s->ext.ech.sent);
+    WPACKET_get_total_written(pkt, &pp_at_end);
+    s->ext.ech.sent_len = pp_at_end - pp_at_start;
+    s->ext.ech.sent = OPENSSL_malloc(s->ext.ech.sent_len);
+    if (s->ext.ech.sent == NULL)
+        return 0;
```

Comment:
Why no SSLfatal() here?

Reply:
Good point.  I had 'em at one point but took 'em out on the false assumption that `OPENSSL_malloc()` raises an error, but I guess that's not the same as an `SSLfatal()` - will add variously